feat(recording) - add frontend support for recording consent

Antreesy opened the pull request:
### ☑️ Resolves

* Fix #10409
* Solution is based on MVP + V2 description (https://github.com/nextcloud/spreed/issues/10348#issuecomment-1725333919)
  * Option to require consent globally in Admin settings
  * When consent is required `1`:
    * Enforce the media settings dialog
    * Show a checkbox that needs ticking before moderator/user can start or join call
    * Moderator setting is visible but disabled
  * When consent is optional `2`:
    * Moderator setting is enabled with "Off" and "Required"
    * Disabled, while call is currently on
  * If user accidentally joined without consent, will be blocked by API
* Increase preview size in MediaSettings, align all content elements up to its width

## 🖌️ UI Checklist

### 🖼️ Screenshots / Screencasts

#### Admin settings
![image](https://github.com/nextcloud/spreed/assets/93392545/5d1bfb75-f407-4474-a76c-12f43a419a62)

#### Moderator settings

| Off (hidden) | Required (disabled) | Configurable V2 |
|---|------------|----------|
| ![image](https://github.com/nextcloud/spreed/assets/93392545/0d0557c4-0ef5-42ae-8c0b-1c781490a94a) | ![image](https://github.com/nextcloud/spreed/assets/93392545/b0786ffc-0a6f-47c7-95b7-d3f255fa6dd2) | ![image](https://github.com/nextcloud/spreed/assets/93392545/1336e343-86a7-4efd-9d5e-7848037bdbd5)

#### User view (Media settings)

| Off (hidden) | Required (not checked) | Required (checked) |
|---|------------|----------|
| ![image](https://github.com/nextcloud/spreed/assets/93392545/f6d67c72-d7a0-4739-b2d3-e86025097e57) | ![image](https://github.com/nextcloud/spreed/assets/93392545/401e51a8-cd3d-46ec-8320-55eb0f61bbb3) | ![image](https://github.com/nextcloud/spreed/assets/93392545/bb846010-33c6-4dc9-8906-dba3fe108774)

### 🚧 Tasks

- [ ] Adjust wording / styles
- [ ] Test with recording package
- [ ] Test frontend (enter in AdminSettings -> Recording backend credentials for working signaling server)

### 🏁 Checklist

- [x] 🌏 Tested with Chrome, Firefox and Safari or should not be risky to browser differences
- [x] 🖌️ Design was reviewed, approved or inspired by the design team
- [x] ⛑️ Tests are included or not possible
- [ ] 📗 User documentation in https://github.com/nextcloud/documentation/tree/master/user_manual/talk has been updated or is not required

Antreesy commented:
cc @marcoambrosini for design review and wording for MVP + V2 solution
cc @danxuliu as it touches signaling, to ensure it won't break anything

nickvergessen commented on src/components/AdminSettings/RecordingServers.vue:
```diff
@@ -73,36 +96,55 @@ import debounce from 'debounce'
+const recordingConsentOptions = [
+	{ id: CALL.RECORDING_CONSENT.OFF, label: t('spreed', 'Off') },
+	{ id: CALL.RECORDING_CONSENT.REQUIRED, label: t('spreed', 'Enabled') },
+	{ id: CALL.RECORDING_CONSENT.OPTIONAL, label: t('spreed', 'Optional') },
```
Hmm it's not "optional" consent, but basically "on conversation level"
Maybe we find  a better wording ;P

DorraJaouad replied:
"Configurable" ?

nickvergessen commented on src/components/ConversationSettings/RecordingConsentSettings.vue:
```diff
@@ -0,0 +1,133 @@
+		<label for="recording-consent_input" class="recording-consent__label">
+			{{ t('spreed', 'Require recording consent before joining call in this conversation') }}
+		</label>
+		<NcSelect v-model="recordingConsentSelected"
```
Don't show the select when it's enforced on admin level

Antreesy replied:
Should we do it the same as for Expiration settings? https://github.com/nextcloud/spreed/pull/10491

nickvergessen replied:
Sounds good.

Antreesy commented on src/components/MediaSettings/MediaSettings.vue:
```diff
@@ -130,6 +130,13 @@
 			</NcNoteCard>
+			<!-- Recording consent warning -->
+			<NcNoteCard v-if="isRecordingConsentRequired" type="warning">
+				<p>{{ t('spreed', 'Your consent to be recorded is required before joining the call.') }}</p>
+				<NcCheckboxRadioSwitch :checked.sync="recordingConsentGiven">
+					{{ t('spreed', 'Give a consent to be recorded') }}
+				</NcCheckboxRadioSwitch>
```
Only if you participating as a speaker, so could be either voice/video/screen, or any combination

nickvergessen replied:
yeah, but permissions could be given later, so we don't know yet.
But that is why I would love to see a more neutral wording

marcoambrosini reviewed (commented):
The wording seems ok, I'd style the component a bit differently, similar to the quote component:
- border instead of background (better text contrast)
- Rounded corners
- Rounded warning bar on the left
- No warning icon
- Same width as the video component above

Quick sketch: 
![IMG_8921 2](https://github.com/nextcloud/spreed/assets/26852655/dd903d16-61b9-48df-a151-07893afff458)

nickvergessen commented:
> I'd style the component a bit differently

It's the wode spread NcNoteCard component. If that should be restyled, I'd should be done upstream I guess.
The usecase is exactly what the component was designed for 🙈

Antreesy commented:
@marcoambrosini I'd agree with width aligning, but there are also other components wider than Video preview:
Device selectors:
![image](https://github.com/nextcloud/spreed/assets/93392545/7ded2bde-8bc2-4131-9504-7cb45fb41bec)
Background selectors:
![image](https://github.com/nextcloud/spreed/assets/93392545/dd413e94-4fb3-464c-af18-7e5836890c9b)

Maybe, we should increase a preview size instead?

danxuliu reviewed (changes requested):
There are two cases in which joining a call fails if recording consent is enabled: [switching between breakout rooms](https://github.com/nextcloud/spreed/blob/d833dd3a632eb8f57a63be771eca971e2987e722/src/App.vue#L347-L351) and [requesting the password of a share](https://github.com/nextcloud/spreed/blob/33c82ec006647582dcf33af4023d5959152d7ce4/src/PublicShareAuthSidebar.vue#L160-L164) (video verification). In those cases the recording consent is not asked, so the API prevents the user from joining the call.

## How to test (scenario 1):
- Open Talk settings
- Enable recording consent for all calls
- Create a new group conversation
- Add another user
- Set up breakout rooms with 2 rooms
- Start a call in that conversation; as it is the first time that a call is started the media settings screen will be shown
- Give recording consent
- Start the actual call
- In a private window, log in as the other user
- Join the conversation
- Join the call; as it is the first time that a call is started the media settings screen will be shown
- Give recording consent
- Join the actual call
- In the original window, start the breakout rooms session

### Expected result
The user switches to the call in the breakout room (I would say that asking for recording consent again would be overkill and that the recording consent from the parent room should be reused; maybe this could be even done in the API itself)

### Actual result
The user switches to the call in the breakout room, but is immediately kicked out

## How to test (scenario 2):
- Open Talk settings
- Enable recording consent for all calls
- Open the Files app
- Create a link share
- Protect the share with a password
- Enable video verification
- Copy the link
- In a private window, open the link
- Request the password
- Set a name for the guest

### Expected result
The media settings screen (or just a recording consent dialog, but I guess it would be better to reuse the media settings screen) is shown to give recording consent

### Actual result
The user seems to join the call but is immediately kicked out

Independently of that, [the media settings screen should be shown](https://github.com/nextcloud/spreed/blob/f3cd81578ef3c1409701bed95e1138b0b2c6204d/src/components/TopBar/CallButton.vue#L339-L340) even if the user unchecked _Always show preview for this conversation_ when recording consent needs to be given to join a call:
## How to test
- Open Talk settings
- Enable recording consent for all calls, or make it optional and enable it in a specific call
- Create a new conversation, and if recording consent is optional, enable it
- Start a call in that conversation; as it is the first time that a call is started the media settings screen will be shown
- Uncheck _Always show preview for this conversation_
- Give recording consent
- Start the actual call
- Leave the call
- Reload the page
- Start a call

### Expected result
The media settings screen is shown to give recording consent

### Actual result
The user seems to join the call but is immediately kicked out

nickvergessen commented:
Case 1 (sending to breakout room) I guess we can fix on the API level by ignoring the setting there and basically allow "sending users" there. However when stopping breakout rooms they would join the parent again in which case we can not detect on API level if that is happening, so we would require the client to send the consent flag there again.

nickvergessen commented:
> However when stopping breakout rooms they would join the parent again in which case we can not detect on API level if that is happening, so we would require the client to send the consent flag there again.

But that is also a bit annoying, because it will be listed as another consent entry, but I guess it's the best step we have atm

nickvergessen commented:
The reconnection when your permissions are modified is also affected?

danxuliu commented:
> The reconnection when your permissions are modified is also affected?

Good catch. Yes, reconnections do not work because the consent is not sent again. The `silent` parameter is not taken into account either, I will check _tomorrow_ as that might need a separate fix to be able to backport it (and therefore this pull request would need a rebase on that fix to be able to also fix renegotiations when consent is required).

marcoambrosini reviewed (dismissed):
@Antreesy I think the use of selects here is not appropriate.
I think that the admin settings should be radio buttons, with an explanation always shown and not in a warning badge.
On the other hand, for the moderators on a per-conversation basis, we should use a toggle.

danxuliu commented:
> The `silent` parameter is not taken into account either, I will check _tomorrow_ as that might need a separate fix to be able to backport it (and therefore this pull request would need a rebase on that fix to be able to also fix renegotiations when consent is required).

Fix for `silent` parameter is in https://github.com/nextcloud/spreed/pull/10688 The second commit can be used as a reference on what to change (after rebasing) to send the consent again on reconnections.

Antreesy commented:
Rebased onto master, change interactive elements and wording
New screenshots attached, review appreciated =)

P.S. only UI changes, logic from Daniel is not touched yet, waiting for 10688 to be merged

Antreesy commented:
Rebase onto master, fix consent usage for App, FilesSidebarChatView, PublicShareSidebar

PublicShareAuthSidebar needs some adjustments to show MediaSettings before joining the call

DorraJaouad commented on src/views/FilesSidebarChatView.vue:
```diff
@@ -23,7 +23,7 @@
 		<CallButton class="call-button" />
 		<ChatView />
-		<MediaSettings :initialize-on-mounted="false" />
+		<MediaSettings :initialize-on-mounted="false" :recording-consent-given.sync="recordingConsentGiven" />
```
So why are you adding `recordingConsentGiven` here ?

Antreesy replied:
to be able to connect from files sidebar also, if consent is set to 1 (applied to all calls on instance)
consent may be reused when reconnecting, so should not be stored in MediaSettings component. Also don't want to get a new store for it

Antreesy replied:
For "configurable" mode, consent will be switched off for new conversations like this, so not a problem
For "required" mode, I'd say that it's a pre-emptive measure (to not come back to it when we add owners/moderators to such conversations)

DorraJaouad replied:
To keep this pre-emptive measure, consent should not be mandatory atm as this might be a GDPR violation https://gdpr.eu/gdpr-consent-requirements/ (see "Consent must be freely given" section)  
> According to [Recital 42](https://gdpr.eu/Recital-42-Burden-of-proof-and-requirements-for-consent/), “Consent should not be regarded as freely given if the data subject has no genuine or free choice or is unable to refuse or withdraw consent without detriment.” 
The one exception is if you need some piece of data from someone to provide them with your service. For example, you may need their credit card information to process a transaction or their mailing address to ship a product.


For those participants who do not wish to consent yet, they want to join the call, you are mandating their consent despite that it does not fall under the one exception because you don't need their consent in this scenario.

Suggested alternative : An opt-in choice to consent or decline without impacting their ability to join the call

nickvergessen replied:
> For those participants who do not wish to consent yet, they want to join the call, you are mandating their consent despite that it does not fall under the one exception because you don't need their consent in this scenario.

This is a customer requirement atm. It is not possible to join without giving consent on API level and this was approved by the customer.

> Suggested alternative : An opt-in choice to consent or decline without impacting their ability to join the call

This can be looked at as a potential follow up, but has quite a long trail (settings needs to be stored on a session based level, recording backend needs access to the user list and filter out peoples audio/video where consent was off, etc). Far out of scope for now, so this is what it will be `¯\_(ツ)_/¯` It's not getting worse then before where the recording was always done without even asking.

nickvergessen replied:
> For those participants who do not wish to consent yet, they want to join the call, you are mandating their consent despite that it does not fall under the one exception because you don't need their consent in this scenario.

That being said, the user can still disable the microphone/camera and will not be recorded. So could say that this is the users choice here and we are basically good again.

Antreesy replied:
It's clear that you're talking about file sharing conversations.
Although there are not any owners / moderators to be able to start the recording, technically it's possible (and we won't return to this code again)
Besides, it's possible to promote user with `occ talk:room:promote [token] [use]`

DorraJaouad replied:
Thanks for sharing that @Antreesy . In that case, can end this legal concerns discussion 🧑🏻‍⚖️for now.

Antreesy commented on src/components/ConversationSettings/RecordingConsentSettings.vue:
```diff
@@ -0,0 +1,123 @@
+		<p v-else-if="isGlobalConsent">
+			{{ t('spreed', 'Recording consent is required for all calls') }}
+		</p>
+		<p v-else>
+			{{ summaryLabel }}
+		</p>
```
Yes, should be in Setup Overview when rebase on top of #10647

Antreesy commented on src/components/MediaSettings/MediaSettings.vue:
```diff
@@ -131,6 +131,21 @@
+				<p>
+					{{ t('spreed', 'The recording might include your voice, video from camera, and screen share. Your consent is required before joining the call.') }}
+				</p>
+				<NcCheckboxRadioSwitch :checked="recordingConsentGiven" @update:checked="setRecordingConsentGiven">
+					{{ t('spreed', 'Give consent to the recording of this call') }}
+				</NcCheckboxRadioSwitch>
```
or rewrite color

Antreesy commented on src/components/ConversationSettings/RecordingConsentSettings.vue:
```diff
@@ -0,0 +1,123 @@
+			default: null,
+		},
+
+		canFullModerate: {
+			type: Boolean,
+			default: true,
+		},
```
needed for Setup overview

Antreesy commented on src/components/MediaSettings/MediaSettings.vue:
```diff
@@ -591,8 +622,8 @@ export default {
 		background-color: var(--color-loading-dark);
-		height: 300px;
-		width: 400px;
+		height: 345px;
+		width: 460px;
```
It was aligned. Maybe, depends on whether scrollbar is visible

Antreesy commented on src/components/MediaSettings/MediaSettings.vue:
```diff
@@ -131,6 +131,21 @@
 			</NcCheckboxRadioSwitch>
 
+			<!-- Recording consent warning -->
+			<NcNoteCard v-if="!isInCall && isRecordingConsentRequired" type="warning">
```
1. It could be another moderator
2. Make sense for me to sync checkboxes with `isRecordingFromStart` and `recordingConsentGiven`

danxuliu reviewed (dismissed):
The media settings screen is not shown if the user unchecked _Always show preview for this conversation_ when recording consent needs to be given. Please refer to the last _How to test_ section in https://github.com/nextcloud/spreed/pull/10633#pullrequestreview-1668480416

Independently of that, the consent sometimes is sent again and sometimes it is not when using breakout rooms (first _How to test_ section in https://github.com/nextcloud/spreed/pull/10633#pullrequestreview-1668480416). I have not been able to find a pattern, maybe it is related to fetching the room information [to compare the parent room when switching to a different breakout room](https://github.com/nextcloud/spreed/pull/10633/commits/690c423a2c97ce2ef24281352de70701339e1af3#diff-7a7a37b12ee1265d7e27577ec286120d2cbc0940908635e264a2be44ccb9a8a0R232-R237) to clean or not to clean the consent, but I do not know.

When requesting the password of a share it might be good to leave the room if the media settings screen is closed without joining the call. Leaving the call or the conversation should update the sidebar and remove the call view and the chat and [show _This conversation has ended_](https://github.com/nextcloud/spreed/blob/33c82ec006647582dcf33af4023d5959152d7ce4/src/PublicShareAuthSidebar.vue#L28), although it seems that was lost at some point :thinking: As leaving the call does not currently do that I guess it would be fine to leave the conversation when the media settings is closed in a follow up once the existing code was also fixed.

To end in a bright note, the consent is now sent again in forced reconnections :-)

Antreesy commented:
> ...if the user unchecked _Always show preview for this conversation_ 

I might have lost these code lines... Moved logic from components to the store, to sync it properly between MediaSettings, ConversationSettingsDialog and CallButton (tests included) :green_circle: 


> ...maybe it is related to fetching the room information

I have relied on breakoutRoomsStore, but it works differently for moderators and non-moderators. Fixed :green_circle: 

> Leaving the call or the conversation should update the sidebar and remove the call view and the chat and [show _This conversation has ended_]

with internal signling only, we fetch conversation within an interval, and show context you mentioned. but with HPB it wait for signaling message. Don't know the logic here (cc @nickvergessen), but it seems like when the share owner leaves conversation, it's removed on server side. So, we may track signaling message, that owner left the room, and fetch the conversation within some timeout => and if it was deleted, all should go back to normal => will raise a follow-up issue

DorraJaouad reviewed (commented):
In case the consent is configurable on conversation level, I noticed when a moderator switches on the consent requirement while a participant has the conversation open ( on a chat view), it's not updated for them thus there is no consent requested when joining the call.

nickvergessen commented:
> In case the consent is configurable on conversation level, I noticed when a moderator switches on the consent requirement while a participant has the conversation open ( on a chat view), it's not updated for them thus there is no consent requested when joining the call.

Missing an update trigger via the HPB then, so it's a backend issue.

nickvergessen commented on src/store/conversationsStore.js:
```diff
@@ -565,6 +566,18 @@ const actions = {
 		commit('addConversation', conversation)
 	},
 
+	async setRecordingConsent({ commit, getters }, { token, state }) {
```
Fixed in https://github.com/nextcloud/spreed/pull/10727

nickvergessen commented:
> In case the consent is configurable on conversation level, I noticed when a moderator switches on the consent requirement while a participant has the conversation open ( on a chat view), it's not updated for them thus there is no consent requested when joining the call.

Also fixed in https://github.com/nextcloud/spreed/pull/10727 now

Antreesy commented:
Rebased onto master, should work for breakout rooms, and fetch conversation on change
Consent is kept for main view and sidebar integrations (turns out it's possible to promote people to moderators there with occ command at least)

DorraJaouad reviewed (approved):
Breakout rooms session should be closed when the moderator wants to toggle the consent requirement in the conversation settings . It renders Bad request otherwise ( maybe add a hint in toast ? )

Overall, nothing is blocking  🦅 .

danxuliu reviewed (commented):
(I did not see that the review was dismissed before actually testing the pull request :facepalm:)

Tested and works :+1: 

A minor detail: when the device selector is shown in _request password_/_video verification_ clicking on _Always show preview for this conversation_ or in the menu to start a silent call has no effect (the checkbox is not modified, and the menu is not opened).

Nevertheless, those options should not be even visible, because every time a password is requested again a new room will be created, so there is no point in remembering _Always show preview for this conversation_, and the call should not be silently started, as the owner of the file should be notified that a password is being requested.
